testsuite/tests/typing-objects-bugs/pr4018_bad.compilers.reference

```diff
@@ -16,26 +16,23 @@ Error: This type entity = < destroy_subject : id subject; entity_id : id >
          'a entity_container =
            < add_entity : (< destroy_subject : < add_observer : 'a
                                                                 entity_container ->
-                                                                'e;
-                                                 .. >;
+                                                                'f;
+                                                 .. >
+                                               as 'e;
                              .. >
                            as 'd) ->
-                          'e;
+                          'f;
              notify : 'd -> id -> unit > 
-       Type < destroy_subject : id subject; entity_id : id >
-       is not compatible with type
-         entity = < destroy_subject : id subject; entity_id : id > 
+       Type entity = < destroy_subject : id subject; entity_id : id >
+       is not compatible with type < destroy_subject : 'e; .. > as 'd 
        Type
-         < add_observer : (id subject, id) observer -> unit;
-           notify_observers : id -> unit >
-       is not compatible with type
          id subject =
            < add_observer : (id subject, id) observer -> unit;
-             notify_observers : id -> unit > 
-       Type < add_entity : 'd -> 'e; notify : 'd -> id -> unit >
+             notify_observers : id -> unit >
+       is not compatible with type
+         < add_observer : 'a entity_container -> 'f; .. > as 'e 
+       Type (id subject, id) observer = < notify : id subject -> id -> unit >
        is not compatible with type
          'a entity_container =
-           < add_entity : 'd -> 'e; notify : 'd -> id -> unit > 
-       Type < notify : id subject -> id -> unit > is not compatible with type
-         (id subject, id) observer = < notify : id subject -> id -> unit > 
-       Types for method add_entity are incompatible
+           < add_entity : 'd -> 'f; notify : 'd -> id -> unit > 
+       The first object type has no method add_entity
```

Comment:
I find the error messages here fairly hard to review. Before the patch, the message is very strange (it starts by saying that two visually identical types are incompatibe). After the patch, some of the strageness is removed, but the error message remains unreadable and incomprehensible. I guess this is unavoidable (without substantially more work) and I am not suggesting/requesting any additional improvement to the error-message logic, but maybe this could be converted to an expect-test, to at least make the test cases more readable?

Reply:
This original error message here is indeed extremely confusing because it lost track of which types were the actual and the expected due to a double expansion of the trace. In the first two suberrors, the error messages is in fact comparing the actual type with itself then the expected type with itself.